### ISSUE TITLE
running "mvn docker:build" can fail now on the latest OpenShift 1.1.1

### ISSUE THREAD

**Issue by jstrachan** [fabric8io/fabric8#5655]
since the upgrade I get this fairly often:
```
[ERROR] DOCKER> conflict: unable to delete 139581d35141 (cannot be forced) - image is being used by running container ff632292ed06 (Conflict: 409)
```
as the 2nd time you rebuild a SNAPSHOT version of a docker image docker barfs that you are rebuilding the image thats already in use.

This makes sense; as you're changing an image thats being used by kubernetes.

Maybe we need to generate a new timestamped image name each time? Then ideally, we'd GC away all previous timestamped images that are no longer in use? e.g.

* do a regular `mvn docker:build fabric8:apply` and the RC starts up
* repeat `mvn docker:build fabric8:apply`, creates a new image and the new RC starts; killing the old pod
* repeat `mvn docker:build fabric8:apply`, this time the build from 1 is no longer used so we can clean that up, then creates a new image and the new RC starts; killing the old pod

**Comment by rhuss** [comment]
For a short term solution (without cleaning up) you can use `<cleanup>false</cleanup>` in the `<build>` configuration section which will avoid this step. 

I wonder whether the 'garbage collection' shouldn't be done completely outside from the plugin (e.g. by some cron jobs).

**Comment by jstrachan** [comment]
I've been hitting this a fair bit lately; the work around seems to be:

* scale the RC to zero
* run `docker rmi -f fabric8/someimage:2.x-SNAPSHOT

which is a little painful. Sometimes there's no error and the build happens; but openshift totally ignores the newly built image (which adds even more confusion)!

**Comment by jstrachan** [comment]
am wondering if for snapshot images we should just generate new labels (like maven deploy of snaphots does). The only issue we'd need to make sure is we use the same label in the docker image and in the generated kubernetes.yaml/json

**Comment by rhuss** [comment]
@jstrachan the problem with not being able to remove has been somewhat fixed in the latest d-m-p version 0.14.2  

There are now three mode for `<cleanup>`:

* **none** : The old image gets not deleted
* **remove** : Remove the old image and fail if not possible
* **try** : Try to remove the old image, and only print a warning if not possible.

The default now is `try` (was: `remove`), so you shouldn't run in an error anymore (but get a warning instead)

----

This might not solve the issue that OpenShift doesnt pick up a new image. With `label` you mean an image tag, right ? (so that we have a e.g. a timestamped image tag which is refered in the JSON files. For building the image you already can add extra labels from Maven props, so that should to the trick. Now the JSON generation needs to use this timestamped label

Does OpenShift ignore the image also when it is tagged with `latest` ? I think thats the OS equivalent to `SNAPSHOT` so it might be worse a try to use this in the image name instead of SNAPSHOT.

**Comment by jstrachan** [comment]
@rhuss yeah, without scaling down an RC you can't even remove the image on OpenShift (even by force). So by `label` I was meaning the docker image label.

Its definitely worth an experiment using `latest` in the kubernetes JSON; then using an actual timestamp for the docker label and seeing if that fixes it!
